url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/Triangle3D.java

```diff
@@ -0,0 +1,3211 @@
+	 * @param numOfAttributes
+	 *            maximum number of attributes which are or will be provided
+	 * @param attributesArray
+	 *            2D String Array holding attributes and values of maximum
+	 *            length of numOfAttributes
+	 */
+	public Triangle3D(Point3D point1, Point3D point2, Point3D point3,
+			ScalarOperator sop, int numOfAttributes) {
+
+		this(point1, point2, point3, sop);
+
+		if (numOfAttributes > 0) {
+			initNumOfAttributes(numOfAttributes);
+		}
+
+	}
+
+	/**
+	 * Constructor<br>
+	 * The given point must not intersect with the segment.
+	 * 
+	 * @param point
+	 *            Point3D
+	 * @param seg
+	 *            Segment3D
+	 * @param sop
+	 *            ScalarOperator, needed for validation. If ScalarOperator is
+	 *            <code>null</code>, no validation will occur.
+	 * @param numOfAttributes
+	 *            maximum number of attributes which are or will be provided
+	 * @param attributesArray
+	 *            2D String Array holding attributes and values of maximum
+	 *            length of numOfAttributes
+	 * 
+	 */
+	public Triangle3D(Point3D point, Segment3D seg, ScalarOperator sop,
+			int numOfAttributes) {
+
+		this(point, seg, sop);
+
+		if (numOfAttributes > 0) {
+			initNumOfAttributes(numOfAttributes);
+		}
+	}
+
+	
+//	Constructors without attributes	
+	
+	/**
+	 * Constructor.
+	 * 
+	 * @param pts
+	 *            Point3D array
+	 * @param sop
+	 *            ScalarOperator, needed for validation. If ScalarOperator is
+	 *            <code>null</code>, no validation will occur
+	 * @throws ValidationException
+	 *             - signals inappropriate parameters.<br>
+	 * @throws IllegalArgumentException
+	 *             - if an attempt is made to construct a Triangle3D from a
+	 *             point array whose length is not 3 or the validation of the
+	 *             constructed Triangle3D fails.
+	 * @throws ArithmeticException
+	 *             - if norm equals zero in epsilon range. This exception
+	 *             originates in the method normalize(ScalarOperator) of the
+	 *             class Vector3D.
+	 */
+	public Triangle3D(Point3D[] pts, ScalarOperator sop)
+			throws IllegalArgumentException {
+		if (pts == null || pts.length != 3)
+			throw new IllegalArgumentException(
+					Db3dSimpleResourceBundle
+							.getString("db3d.geom.trionlythree"));
+
+		this.zero = pts[0];
+		this.one = pts[1];
+		this.two = pts[2];
+		this.normvec = null;
+		this.lines = null;
+
+		// validate
+		if (sop != null) {
+			if (!isValid(sop))
+				throw new IllegalArgumentException(
+						Db3dSimpleResourceBundle
+								.getString("db3d.geom.argnotval"));
+			if (!isRegular(sop))
+				throw new IllegalArgumentException(
+						Db3dSimpleResourceBundle
+								.getString("db3d.geom.argnotreg"));
+			if (!isBeautiful(sop))
+				throw new IllegalArgumentException(
+						Db3dSimpleResourceBundle
+								.getString("db3d.geom.argnotbeau"));
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 450:463, 496:509.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.